File: app/src/main/java/com/shahin/movieapp/view/DetailMovieFragment.java
package com.shahin.movieapp.view;

import android.os.Bundle;

import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.material.snackbar.Snackbar;
import com.sergivonavi.materialbanner.Banner;
import com.shahin.movieapp.R;
import com.shahin.movieapp.databinding.FragmentDetailMovieBinding;
import com.shahin.movieapp.model.Movie;
import com.shahin.movieapp.viewModel.MainViewModel;
import com.squareup.picasso.Picasso;

public class DetailMovieFragment extends Fragment {

    public static final String ARG_MOVIE = "ArgMovie";
    private FragmentDetailMovieBinding mMovieBinding;
    private MainViewModel mMainViewModel;
    private String mImdbId;
    private Banner mBanner;
    private boolean flagOffLine = false;

    public DetailMovieFragment() {
        // Required empty public constructor
    }

    public static DetailMovieFragment newInstance(String imdbId) {
        DetailMovieFragment fragment = new DetailMovieFragment();
        Bundle args = new Bundle();
        args.putString(ARG_MOVIE, imdbId);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mImdbId = getArguments().getString(ARG_MOVIE);
        }

        mMainViewModel = new ViewModelProvider(this).get(MainViewModel.class);

        if (mMainViewModel.isOnline(getContext())) {
            sendRequest();
        } else {
            flagOffLine = true;
        }

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        mMovieBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_detail_movie, container, false);

        if (flagOffLine) {
            Movie movieDB = mMainViewModel.getMovieFromDB(mImdbId);
            if (movieDB != null) {
                if (movieDB.getRuntime() != null && !movieDB.getRuntime().equals("")) {
                    initUI(movieDB);
                } else {
                    initPic(movieDB);
                    initBanner();
                }
            }
        }

        return mMovieBinding.getRoot();
    }

    public void initBanner() {
        mBanner = new Banner.Builder(getContext())
                .setParent(mMovieBinding.llBannerDetail)
                .setIcon(R.drawable.ic_alart)
                .setMessage("You have lost connection to the Internet. This app is offline.")
                .setLeftButton("Dismiss", banner -> {
                    mBanner.dismiss();

                })
                .setRightButton("Turn on wifi", banner -> {
                    if (mMainViewModel.isOnline(getContext())) {
                        mBanner.setVisibility(View.GONE);
                        mBanner.dismiss();
                        sendRequest();
                        visibleCompo(true);
                    } else {
                        mBanner.show();
                        mBanner.setVisibility(View.VISIBLE);
                    }
                })
                .create();

        mBanner.show();
        mBanner.setVisibility(View.VISIBLE);
    }

    private void sendRequest() {
        mMainViewModel.getMovie(mImdbId).observe(this, movie -> {
            initUI(movie);
            mMainViewModel.addMovieToDB(movie);
        });
    }

    private void initUI(Movie movie) {
        mMovieBinding.tvTitleDetailMovie.setText(movie.getTitle());
        mMovieBinding.tvTimeDetailMovie.setText(movie.getRuntime());
        mMovieBinding.tvYearDetailMovie.setText(movie.getYear());
        mMovieBinding.tvImdbRateDetailMovie.setText(movie.getImdbRating());
        mMovieBinding.tvMetascoreDetailMovie.setText("% " + movie.getMetascore());
        mMovieBinding.tvGenreDetailMovie.setText(movie.getGenre());
        mMovieBinding.tvCountryDetailMovie.setText(movie.getCountry());
        mMovieBinding.tvLanguageDetailMovie.setText(movie.getLanguage());
        mMovieBinding.tvDirectorDetailMovie.setText(movie.getDirector());
        mMovieBinding.tvWriterDetailMovie.setText(movie.getWriter());
        mMovieBinding.tvActorsDetailMovie.setText(movie.getActors());
        mMovieBinding.tvAwardDetailMovie.setText(movie.getAwards());
        mMovieBinding.tvPlotDetailMovie.setText(movie.getPlot());

        Picasso.with(getContext()).load(movie.getPoster()).fit().into(mMovieBinding.ivPicDetailMovie);
        Picasso.with(getContext()).load(movie.getPoster()).into(mMovieBinding.ivSmallPicDetailMovie);

    }

    private void initPic(Movie movie) {
        Picasso.with(getContext()).load(movie.getPoster()).fit().into(mMovieBinding.ivPicDetailMovie);
        Picasso.with(getContext()).load(movie.getPoster()).into(mMovieBinding.ivSmallPicDetailMovie);
        visibleCompo(false);
    }

    private void visibleCompo(boolean isVisible){
        if(isVisible){
            mMovieBinding.linearLayoutCompat4.setVisibility(View.VISIBLE);
            mMovieBinding.linearLayoutCompat2.setVisibility(View.VISIBLE);
            mMovieBinding.linearLayoutCompat.setVisibility(View.VISIBLE);
            mMovieBinding.ivShowFilmDetailMovie.setVisibility(View.VISIBLE);
        }else {
            mMovieBinding.linearLayoutCompat4.setVisibility(View.INVISIBLE);
            mMovieBinding.linearLayoutCompat2.setVisibility(View.INVISIBLE);
            mMovieBinding.linearLayoutCompat.setVisibility(View.INVISIBLE);
            mMovieBinding.ivShowFilmDetailMovie.setVisibility(View.INVISIBLE);
        }
    }

}

File: app/src/main/java/com/shahin/movieapp/ui/moviedetail/DetailMovieFragment.kt
package com.shahin.movieapp.ui.moviedetail

import android.content.Context
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.navArgs
import com.shahin.movieapp.R
import com.shahin.movieapp.databinding.FragmentDetailMovieBinding
import com.shahin.movieapp.di.ViewModelFactory
import com.shahin.movieapp.model.Movie
import com.shahin.movieapp.ui.MainActivity
import com.shahin.movieapp.ui.utiles.bannerNetwork
import com.shahin.movieapp.ui.utiles.glideLoad
import javax.inject.Inject

class DetailMovieFragment : Fragment() {

    private lateinit var binding: FragmentDetailMovieBinding

    @Inject
    lateinit var viewModelFactory : ViewModelFactory
    private val viewModel by viewModels<DetailMovieViewModel> { viewModelFactory }

    private val navArgument : DetailMovieFragmentArgs by navArgs()

    override fun onAttach(context: Context) {
        super.onAttach(context)
        (requireActivity() as MainActivity).mainActivitySubComponent.inject(this)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_detail_movie, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewModel.getMovie(navArgument.imdbId).observe(viewLifecycleOwner) { movie ->
            movie?.let {
                initUI(movie)
                initPic(movie)
            }
        }

        viewModel.error.observe(viewLifecycleOwner){
            if (binding.tvTitleDetailMovie.text.isNullOrEmpty())
            requireContext().bannerNetwork(binding.llBannerDetail) { viewModel.isOnline(requireContext()) }
        }
    }

    private fun initUI(movie: Movie) {
        binding.tvTitleDetailMovie.text = movie.title
        binding.tvTimeDetailMovie.text = movie.runtime
        binding.tvYearDetailMovie.text = movie.year
        binding.tvImdbRateDetailMovie.text = movie.imdbRating
        binding.tvMetascoreDetailMovie.text = "% " + movie.metaScore
        binding.tvGenreDetailMovie.text = movie.genre
        binding.tvCountryDetailMovie.text = movie.country
        binding.tvLanguageDetailMovie.text = movie.language
        binding.tvDirectorDetailMovie.text = movie.director
        binding.tvWriterDetailMovie.text = movie.writer
        binding.tvActorsDetailMovie.text = movie.actors
        binding.tvAwardDetailMovie.text = movie.awards
        binding.tvPlotDetailMovie.text = movie.plot

        requireContext().glideLoad(url = movie.poster,imageView = binding.ivPicDetailMovie)
        requireContext().glideLoad(url = movie.poster,imageView = binding.ivSmallPicDetailMovie)
    }

    private fun initPic(movie: Movie) {
        requireContext().glideLoad(url = movie.poster,imageView = binding.ivPicDetailMovie)
        requireContext().glideLoad(url = movie.poster,imageView = binding.ivSmallPicDetailMovie)
        visibleCompo(false)
    }

    private fun visibleCompo(isVisible: Boolean) {
        if (isVisible) {
            binding.linearLayoutCompat4.visibility = View.VISIBLE
            binding.linearLayoutCompat2.visibility = View.VISIBLE
            binding.linearLayoutCompat.visibility = View.VISIBLE
            binding.ivShowFilmDetailMovie.visibility = View.VISIBLE
        } else {
            binding.linearLayoutCompat4.visibility = View.INVISIBLE
            binding.linearLayoutCompat2.visibility = View.INVISIBLE
            binding.linearLayoutCompat.visibility = View.INVISIBLE
            binding.ivShowFilmDetailMovie.visibility = View.INVISIBLE
        }
    }

}